Feature Request: Option to exclude certain file(s) from "openfiles" check

This came up when the error below was encountered in an `astropy.visualization` test. It is suspected that this might be a problem upstream at `matplotlib` and we would want to mark this not unlike `xfail` but we don't want to `xfail` the whole test unconditionally but only when `pytest-openfiles` detected the files being open.

```
==================================== ERRORS ====================================
_______________________ ERROR at teardown of test_units ________________________
item = <Function 'test_units'>, nextitem = <Function 'test_incompatible_units'>
    def pytest_runtest_teardown(item, nextitem):
        # a "skipped" test will not have been called with
        # pytest_runtest_setup, so therefore won't have an
        # "open_files" member
        if (not item.config.getvalue('open_files') or
                not hasattr(item, 'open_files')):
            return
    
        start_open_files = item.open_files
        del item.open_files
    
        open_files = _get_open_file_list()
    
        # This works in tandem with the test_open_file_detection test to
        # ensure that it creates one extra open file.
        if item.name == 'test_open_file_detection':
            assert len(start_open_files) + 1 == len(open_files)
            return
    
        not_closed = set()
        open_files_ignore = item.config.getini('open_files_ignore')
        for filename in open_files:
            ignore = False
    
            for ignored in open_files_ignore:
                if not os.path.isabs(ignored):
                    if os.path.basename(filename) == ignored:
                        ignore = True
                        break
                else:
                    if filename == ignored:
                        ignore = True
                        break
    
            if ignore:
                continue
    
            if filename not in start_open_files:
                not_closed.add(filename)
    
        if len(not_closed):
            msg = ['File(s) not closed:']
            for name in not_closed:
                msg.append('  {0}'.format(name))
>           raise AssertionError('\n'.join(msg))
E           AssertionError: File(s) not closed:
E             /home/travis/miniconda/envs/test/lib/python3.5/site-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSansDisplay.ttf
E             /home/travis/miniconda/envs/test/lib/python3.5/site-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSans.ttf
```

xref astropy/astropy#6942

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1,6 +1,8 @@
 0.2.1 (unreleased)
 ==================
 
+- Add decorator to skip detection of open files for particular tests. [#10]
+
 
 0.2 (2017-12-07)
 ================

--- a/README.rst
+++ b/README.rst
@@ -65,6 +65,21 @@ they reside in the file system::
 In this example, all files named ``output.log`` will be ignored if they are
 found to remain open after a test completes.
 
+It is also possible to ignore open files for particular test cases by
+decorating them with the ``openfiles_ignore`` decorator:
+
+.. code::
+
+    import pytest
+
+    @pytest.mark.openfiles_ignore
+    def test_open_file_and_ignore():
+        """We want to ignore this test when checking for open file handles."""
+
+
+The test function will not be skipped, but any files that are left open by the
+test will be ignored by this plugin.
+
 
 Development Status
 ------------------

--- a/pytest_openfiles/plugin.py
+++ b/pytest_openfiles/plugin.py
@@ -24,6 +24,11 @@ def pytest_addoption(parser):
                   "may be specified by their base name (ignoring their full "
                   "path) or by absolute path", type="args", default=())
 
+def pytest_configure(config):
+
+    config.getini('markers').append(
+        'openfiles_ignore: Indicate that open files should be ignored for this test')
+
 # Open file detection.
 #
 # This works by calling out to psutil to get the list of open files
@@ -53,7 +58,7 @@ def _get_open_file_list():
 def pytest_runtest_setup(item):
     # Store a list of the currently opened files so we can compare
     # against them when the test is done.
-    if item.config.getvalue('open_files'):
+    if item.config.getvalue('open_files') and not item.get_marker('openfiles_ignore'):
         item.open_files = _get_open_file_list()
 
 
@@ -61,8 +66,7 @@ def pytest_runtest_teardown(item, nextitem):
     # a "skipped" test will not have been called with
     # pytest_runtest_setup, so therefore won't have an
     # "open_files" member
-    if (not item.config.getvalue('open_files') or
-            not hasattr(item, 'open_files')):
+    if (not item.config.getvalue('open_files') or not hasattr(item, 'open_files')):
         return
 
     start_open_files = item.open_files

--- a/tests/test_open_file_detection.py
+++ b/tests/test_open_file_detection.py
@@ -1,6 +1,8 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 import os
 
+import pytest
+
 PKG_DATA_DIR = os.path.dirname(__file__)
 
 fd = None
@@ -10,6 +12,10 @@ def test_open_file_detection():
     global fd
     fd = open(os.path.join(PKG_DATA_DIR, 'data/open_file_detection.txt'))
 
+@pytest.mark.openfiles_ignore
+def test_skip_open_file_detection():
+    global fd
+    fd = open(os.path.join(PKG_DATA_DIR, 'data/open_file_detection.txt'))
 
 def teardown():
     if fd is not None:

